### PR TITLE
Per May 2016 TC39 consensus, `Object.getOwnPropertyDescriptors` is now stage 4!

### PR DESCRIPTION
**Pull request by ljharb**
(https://github.com/tc39/proposal-object-getownpropertydescriptors/issues/9)

### CODE REVIEW

File: spec.html [hunk 1]
```diff
@@ -22683,6 +22683,21 @@
...
+        <h1>Object.getOwnPropertyDescriptors ( _O_ )</h1>
+        <p>When the `getOwnPropertyDescriptors` function is called, the following steps are taken:</p>
+        <emu-alg>
+          1. Let _obj_ be ? ToObject(_O_).
+          1. Let _ownKeys_ be ? _obj_.[[OwnPropertyKeys]]().
+          1. Let _descriptors_ be ! ObjectCreate(%ObjectPrototype%).
+          1. Repeat, for each element _key_ of _ownKeys_ in List order,
+            1. Let _desc_ be ? _obj_.[[GetOwnProperty]](_key_).
+            1. Let _descriptor_ be ! FromPropertyDescriptor(_desc_).
+            1. Perform ! CreateDataProperty(_descriptors_, _key_, _descriptor_).
```
Comment (claudepache) [inline spec.html]: Should we run this step also when _descriptor_ is _undefined_? (Could happen if _obj_ is a proxy, for instance.)

Comment (ljharb) [inline spec.html]: I think that's perfectly fine. If a Proxy chooses to expose that it is a Proxy by providing a key in `[[OwnPropertyKeys]]` that does not have a corresponding descriptor in `[[GetOwnProperty]]`, I think that intentional information leak should be propagated.

Comment (WebReflection) [inline spec.html]: what happens if you `Object.getOwnpropertyDescriptor(proxy, key)` and/or
`Reflect.ownKeys(proxyObject)` ?

Comment (ljharb) [inline spec.html]: `Reflect.ownKeys` should return the same as `[[OwnPropertyKeys]]`, and `Object.getOwnPropertyDescriptor` I believe will call `[[GetOwnProperty]]`, so I think you'd get `undefined` for a key that had this inconsistent behavior.

Comment (claudepache) [inline spec.html]: If `Object.defineProperties` accepts **undefined** in place of a property descriptor, it is probably fine for `Object.getOwnPropertyDescriptors` to produce **undefined** in place of a property descriptor, so that the following code will work:
```js
Object.defineProperties(obj2, Object.getOwnPropertyDescriptors(obj1))
```
I’ve checked that `Object.defineProperties` does indeed accept **undefined** in place of a descriptor since ES6 ([see step 5b](https://tc39.github.io/ecma262/#sec-objectdefineproperties)), but the browsers still do not.

Comment (leobalter) [inline spec.html]: > If Object.defineProperties accepts undefined in place of a property descriptor, it is probably fine for Object.getOwnPropertyDescriptors to produce undefined in place of a property descriptor

This is wrong. `defineProperties` and `defineProperty` does not accept a non-object descriptor values, they properly throw a TypeError. There's no update to be taken at test262 wrt these statements.

Comment (leobalter) [inline spec.html]: please, notice that FromPropertyDescriptor is different than ToPropertyDescriptor, where the former accepts undefined returning undefined and the latter throws a TypeError.

FromPropertyDescriptor is already used within `{Object, Reflect}.getOwnPropertyDescriptor` methods and it's reasonable to find it here with Object.getOwnPropertyDescriptors as well.

What can be changed, and that's not breaking anything is removing the `!` signal from `1. Let _descriptor_ be ! FromPropertyDescriptor(_desc_).`, it does not throw in any way.

Comment (claudepache) [inline spec.html]: In case it was not clear enough: For `Object.defineProperties`, the special case of _undefined_ is handled by [step 5b of ObjectDefineProperties()](https://tc39.github.io/ecma262/#sec-objectdefineproperties).

Comment (claudepache) [inline spec.html]: Sorry, I have indeed misread in step 5b of `ObjectDefineProperties()ss:
>  _propDesc_ is not undefined

is not:

> _propDesc_.[[Value]] is not undefined (or something).

Comment (claudepache) [inline spec.html]: So returning to the original issue, we have a problem here: Since `Object.getOwnPropertyDescriptors` may rarely produce **undefined** in place of a property descriptor, the following pattern may rarely fail:

```js
Object.defineProperties(obj2, Object.getOwnPropertyDescriptors(obj1))
```

Not nice.

Comment (ljharb) [inline spec.html]: @claudepache it *should* fail in the case of a badly formed Proxy that returns disagreeing results from `[[OwnPropertyKeys]]` and `[[GetOwnProperty]]` - we shouldn't try to hard to protect developers who go out of their way to break things.

Comment (claudepache) [inline spec.html]: @ljharb I wasn’t thinking of a broken proxy, but of a semi-plausible situation where [[OwnPropertyKeys]] is consistent but a [[GetOwnProperty]] call manage to delete a not-yet-visited property listed by [[OwnPropertyKeys]].

Comment (ljharb) [inline spec.html]: How could that be possible? `[[GetOwnProperty]]` does not invoke a getter (unlike, say, `Object.values`), and as such, it should not be possible for a key to be deleted mid-enumeration.

Comment (claudepache) [inline spec.html]: > How could that be possible? [[GetOwnProperty]] does not invoke a getter (unlike, say, Object.values), and as such, it should not be possible for a key to be deleted mid-enumeration.

??? Just checked in case I missed something obvious: According to step 8 of [9.5.5](https://tc39.github.io/ecma262/#sec-proxy-object-internal-methods-and-internal-slots-getownproperty-p), the [[GetOwnProperty]] internal method of a Proxy may run arbitrary user-provided code (in particular, delete a key).

Comment (ljharb) [inline spec.html]: Sorry for the miscommunication - you'd said you weren't  thinking of a broken Proxy, so I was referring to non-Proxy use cases.

To sum up: if a Proxy breaks invariants by doing inconsistent things inside and between `[[OwnPropertyKeys]]` and `[[GetOwnProperty]]`, then receiving `undefined` as the property descriptor is completely acceptable imo.

Comment (leobalter) [inline spec.html]: > To sum up: if a Proxy breaks invariants by doing inconsistent things inside and between [[OwnPropertyKeys]] and [[GetOwnProperty]], then receiving undefined as the property descriptor is completely acceptable imo.

This sounds reasonable to me.

Comment (bterlson) [inline spec.html]: @claudepache this is a great issue, thanks for finding it :-P

I believe that this API should always return valid descriptors even with a misbehaving proxy. It doesn't seem onerous to check that you don't have an undefined descriptor before adding the property to the result object. But this is a weak preference and I am happy to be overruled.

Comment (domenic) [inline spec.html]: I guess my similarly-weak preference is that it should just pass through what the MOP gives it and not be responsible for normalizing or censoring.

Comment (ljharb) [inline spec.html]: My weak preference is the same as @domenic's - also, this same "issue" already exists with
```js
const obj = {};
Reflect.ownKeys(badProxy).forEach(x => {
  Object.defineProperty(obj, x, Reflect.getOwnPropertyDescriptor(badProxy, x));
});
```

Comment (bterlson) [inline spec.html]: I don't see this as normalizing or censoring. To me, if GetOwnProperty returns undefined, the property doesn't exist. The fact that previously OwnKeys said the property was there is of no consequence - maybe it doesn't now because the proxy is confused, or maybe there were side effects that removed it.

Comment (domenic) [inline spec.html]: > I don't see this as normalizing or censoring. To me, if GetOwnProperty returns undefined, the property doesn't exist. 

I guess this is the core of our disagreement: if GetOwnProperty returns undefined, that to me means that its property descriptor is undefined. That can be true if the property doesn't exist, but it can also be true if the proxy wants to return undefined and the property does exist.

Basically, "what does it mean to exist"? :)

Comment (bterlson) [inline spec.html]: There is no such thing as a property with an undefined property descriptor. Proxies may change their answers about whether a property exists or not, but if a property exists, it has a property descriptor.

**Comment by bterlson** [comment]
So I've been pushed to a stronger opinion that the only conceptual model that makes sense is that a property for which GetOwnProperty returns undefined does not exist in any sense, and that O.gOPDs should not include either non-existent properties (regardless of whether it is missing because of a confused proxy or other side effects) or invalid descriptors (viz. undefined).

I think this is an issue that is worth discussing in committee, though, so I will accept this PR as-is per May committee consensus and @ljharb will make a needs-consensus PR to fix this bug/feature.

**Comment by leobalter** [comment]
👍 @bterlson.

I believe we can solve it with a If step after _desc_ to check it's not undefined.

I agree with @ljharb and @domenic as this looks a normalization, I also believe it solves at least reusability with broken Proxies. This is also a very specific case.

Even though, it doesn't hurt anyone applying this update after a new consensus. I won't object if @ljharb does it.

**Comment by allenwb** [comment]
I don't see how this feature can be at stage 4 when issues like this are still unsettled. Stage 4 means the spec. is complete and final.

**Comment by ljharb** [comment]
As with many of the things that have needed erratum, this can be considered a spec bug.

**Comment by bterlson** [comment]
@allenwb we made a mistake and didn't realize this issue until @claudepache brought it up. We could say that the consensus was given in error and move it back to stage 3, but I thought it would be fine to do a needs-consensus PR as a follow-up (see #593). Would you prefer the former?

**Comment by ljharb** [comment]
It's also worth noting that in the May meeting, I suggested the possibility of it being acceptable for a proposal to move backwards in stages as necessary, and that possibility was loudly and unanimously rejected.

**Comment by allenwb** [comment]
Makes me wonder about what sort of review it went through before being present to the committee for stage 4.

Stage 4 should mean that the spec. is done and can be safely implemented in production.  If we know that it is i not complete and must be revised (and in particular if there is not yet committee consensus on the modified semantics) then it don't meet the stage 4 criteria, even if the committee had previous said it did.

What hard is there in leaving it at stage 3 until this is all resolved.

**Comment by bterlson** [comment]
There is no leaving it at stage 3. It got stage 4 at the last meeting.

**Comment by bterlson** [comment]
Submitted too soon... I am not opposed to saying that if we find bugs like this it goes back to stage 3. It's never been done before so I thought I'd use the existing processes (namely, needs-consensus PR). Probably a good agenda item for the next meeting :-P

**Comment by leobalter** [comment]
> Stage 4 should mean that the spec. is done and can be safely implemented in production.

Agreed but I believe this proposal meets the criteria as presented in this PR. The follow-up PR has an improvement that makes the method more useful.

**Comment by bterlson** [comment]
Yeah that's a good point too. I think it's a bug (probably @allenwb does too), but the champion and others have been arguing that it's working as intended :-P
